src/test/socket/socket/test_socket.rs

```diff
@@ -0,0 +1,501 @@
+) -> Result<(), ()>
+where
+    I: Iterator<Item = &'a (SocketFn, SocketArguments)>,
+{
+    // the set of file descriptors returned
+    let mut used_fds = std::collections::HashSet::new();
+
+    // assume that we haven't closed 0, 1, or 2
+    used_fds.insert(0);
+    used_fds.insert(1);
+    used_fds.insert(2);
+
+    // number of tests that passed, and the total performed
+    let mut num_passed = 0;
+    let mut num_performed = 0;
+
+    // run the tests
+    for (socket_fn, args) in tests {
+        // by default, we don't expect to fail or want to check the errno
+        let mut should_error = false;
+        let mut expected_errno = None;
+
+        // check the failure conditions
+        for cond in error_conditions.iter() {
+            // if the arguments match one of the failure conditions
+            if args.matches(cond) {
+                should_error = true;
+                expected_errno = cond.expected_errno;
+                break;
+            }
+        }
+
+        print!("Testing {}...", socket_fn.format(args));
+
+        num_performed += 1;
+
+        let fd;
+        match check_socket_call(args, *socket_fn, should_error, expected_errno) {
+            // the test passed
+            Ok(result_fd) => {
+                println!(" ✓");
+                fd = result_fd;
+                num_passed += 1;
+            }
+            // the test failed
+            Err(err) => {
+                println!(" ✗ ({})", err);
+                fd = None;
+                // if not summarizing, should exit immediately
+                if !summarize {
+                    return Err(());
+                }
+            }
+        }
+
+        // check for duplicate fds
+        if let Some(fd) = fd {
+            assert!(!used_fds.contains(&fd), "Duplicate fd returned: {}", fd);
+            used_fds.insert(fd);
+        }
+    }
+
+    println!("Passed {}/{}", num_passed, num_performed);
+
+    if num_passed == num_performed {
+        Ok(())
+    } else {
+        Err(())
+    }
+}
+
+fn get_errno() -> i32 {
+    std::io::Error::last_os_error().raw_os_error().unwrap()
+}
+
+fn get_errno_message(errno: i32) -> String {
+    let cstr;
+    unsafe {
+        let error_ptr = libc::strerror(errno);
+        cstr = std::ffi::CStr::from_ptr(error_ptr)
+    }
+    cstr.to_string_lossy().into_owned()
+}
+
+fn check_socket_call(
+    args: &SocketArguments,
+    socket_fn: SocketFn,
+    should_error: bool,
+    expected_errno: Option<libc::c_int>,
+) -> Result<Option<libc::c_int>, String> {
+    // run either socket() or syscall()
+    let rv = match socket_fn {
+        SocketFn::Socket => unsafe {
+            libc::socket(args.domain, args.sock_type | args.flag, args.protocol)
+        },
+        SocketFn::Syscall => unsafe {
+            libc::syscall(
+                libc::SYS_socket,
+                args.domain,
+                args.sock_type | args.flag,
+                args.protocol,
+            ) as i32
```

Comment:
I'm not sure what `as i32` does, but doesn't `syscall` return an i64 type on our supported platforms?

Reply:
`syscall` returns an i64, but calling it with `SYS_socket` should never return greater than 2^31-1. I'll add a check first before downcasting it just to be safe.